peps/pep-0800.rst

```diff
@@ -0,0 +1,352 @@
+so we disallow ``@solid_base`` on ``Protocol`` definitions.
+
+Similarly, the concept of a "solid base" is not meaningful on ``TypedDict`` definitions, as TypedDicts are purely structural types.
+
+Although they receive some special treatment in the type system, ``NamedTuple`` definitions create real nominal classes that can
+have child classes, so it makes sense to allow ``@solid_base`` on them and treat them like regular classes for the purposes
+of the solid base mechanism.
+
+Specification
+=============
+
+A decorator ``@typing.solid_base`` is added to the type system. It may only be used on nominal classes, including ``NamedTuple``
+definitions; it is a type checker error to use the decorator on a function, ``TypedDict`` definition, or ``Protocol`` definition.
+
+We define two properties on (nominal) classes: a class may or may not *be* a solid base, and every class must *have* a valid solid base.
+
+A class is a solid base if it is decorated with ``@typing.solid_base``, or if it contains a non-empty ``__slots__`` definition.
+This includes classes that have ``__slots__`` because of the ``@dataclass(slots=True)`` decorator or
+because of the use of the ``dataclass_transform`` mechanism to add slots.
+The universal base class, ``object``, is also a solid base.
+
+To determine a class's solid base, we look at all of its base classes to determine a set of candidate solid bases. For each base
+that is itself a solid base, the candidate is the base itself; otherwise, it is the base's solid base. If the candidate set contains
+a single solid base, that is the class's solid base. If there are multiple candidates, but one of them is a subclass of all other candidates,
+that class is the solid base. If no such candidate exists, the class does not have a valid solid base, and therefore cannot exist.
+
+Type checkers must check for a valid solid base when checking class definitions, and emit a diagnostic if they encounter a class
+definition that lacks a valid solid base. Type checkers may also use the solid base mechanism to determine whether types are disjoint,
+for example when checking whether a type narrowing construct like ``isinstance()`` results in an unreachable branch.
+
+Example::
+
+    from typing import solid_base, assert_never
+
+    @solid_base
+    class Solid1:
+        pass
+
+    @solid_base
+    class Solid2:
+        pass
+
+    @solid_base
+    class SolidChild(Solid1):
+        pass
+
+    class C1:  # solid base is `object`
+        pass
+
+    # OK: candidate solid bases are `Solid1` and `object`, and `Solid1` is a subclass of `object`.
+    class C2(Solid1, C1):  # solid base is `Solid1`
+        pass
+
+    # OK: candidate solid bases are `SolidChild` and `Solid1`, and `SolidChild` is a subclass of `Solid1`.
+    class C3(SolidChild, Solid1):  # solid base is `SolidChild`
+        pass
+
+    class C4(Solid1, Solid2):  # error: no single solid base
```

Comment:
could possibly spell this out a bit more

```suggestion
    # error: candidate solid bases are `Solid1` and `Solid2`,
    # but `Solid1` is not a subclass of `Solid2` and `Solid2` is not a subclass of `Solid1`
    class C4(Solid1, Solid2):
```

Reply:
Went with a wording closer to what you suggested on the CPython docs PR :)